core/src/main/scala/filodb.core/memstore/OnDemandPagingShard.scala

```diff
@@ -91,19 +91,24 @@ TimeSeriesShard(ref, schemas, storeConfig, quotaSource, shardNum, bufferMemoryMa
     )
     // TODO the below does not return a particular kind of error code. Most likely this would translate to
     // an internal error of FiloDB which is not appropriate for the case
-    require(estDataSize < enforcedLimits.timeSeriesSamplesScannedBytes,
-      s"With match of $numTsPartitions time series, estimate of $estDataSize bytes exceeds limit of " +
+    if (estDataSize > enforcedLimits.timeSeriesSamplesScannedBytes) {
+      val exMessage = s"With match of $numTsPartitions time series, estimate of $estDataSize bytes exceeds limit of " +
         s"${enforcedLimits.timeSeriesSamplesScannedBytes} bytes queried per shard " +
         s"for ${schemas.apply(schemaId).name} schema. " +
         s"Try one or more of these: " +
         s"(a) narrow your query filters to reduce to fewer than the current $numTsPartitions matches " +
-        s"(b) reduce query time range, currently at ${queryDurationMs / 1000 / 60} minutes")
-    require(numTsPartitions < enforcedLimits.timeSeriesScanned,
+        s"(b) reduce query time range, currently at ${queryDurationMs / 1000 / 60} minutes"
+      throw new QueryLimitException(exMessage, qContext.queryId)
```

Comment:
You are changing the exception type from ``IllegalArgumentException`` to ``RuntimeException``.

Reply:
correct but IllegalArgumentException() is a runtime too. For me it's important to treat them differently. "require" will come up as 500 at the end of the day and I want it to be 422.